src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs

```diff
@@ -170,14 +176,28 @@ protected override void AddPendingTask(ComponentState? componentState, Task task
         base.AddPendingTask(componentState, task);
     }
 
+    private void StopRenderer()
+    {
+        _rendererIsStopped = true;
+    }
+
+    protected override void ProcessPendingRender()
+    {
+        if (_rendererIsStopped)
+        {
+            return;
+        }
+        base.ProcessPendingRender();
+    }
+
```

Comment:
It would be good to add an explanatory comment here. Something like:

When the application triggers a `NotFound` event, we continue rendering the current batch, but after that batch, we don't want to process any pending UI update, as we are going to return 404 and throw away the UI updates we've generated so far.